go-controller/pkg/metrics/metrics.go

```diff
@@ -144,6 +172,177 @@ func coverageShowMetricsUpdater(component string) {
+		return nil, fmt.Errorf("failed to get stopwatch/show output for %s (stderr: %s): %w", component, stderr, err)
+	}
+
+	parsedOutput := parseStopwatchShowOutput(stdout)
+	return parsedOutput, err //need to return err, as it could be set in the defer
+}
+
+// parseStopwatchShowOutput returns the number of total samples for each poll loop
+func parseStopwatchShowOutput(output string) map[string]stopwatchStatistics {
+	result := make(map[string]stopwatchStatistics)
+
+	reStatisticsBlock := regexp.MustCompile(`(?m)^Statistics for '(?P<name>.+)'$(?:\n  .*)+`)
+	reTotalSamples := regexp.MustCompile(`(?m)^  Total samples: (?P<totalSamples>\d*)$`)
+	reMaximum := regexp.MustCompile(`(?m)^  Maximum: (?P<max>\d*) msec$`)
+	reMinumum := regexp.MustCompile(`(?m)^  Minimum: (?P<min>\d*) msec$`)
+	rePercentile95th := regexp.MustCompile(`(?m)^  95th percentile: (?P<percentile95th>\d*\.?\d*) msec$`)
+	reShortTermAvg := regexp.MustCompile(`(?m)^  Short term average: (?P<shortTermAvg>\d*\.?\d*) msec$`)
+	reLongTermAvg := regexp.MustCompile(`(?m)^  Long term average: (?P<longTermAvg>\d*\.?\d*) msec$`)
+
+	nameIndex := reStatisticsBlock.SubexpIndex("name")
+	totalSamplesIndex := reTotalSamples.SubexpIndex("totalSamples")
+	maxIndex := reMaximum.SubexpIndex("max")
+	minIndex := reMinumum.SubexpIndex("min")
+	percentile95thIndex := rePercentile95th.SubexpIndex("percentile95th")
+	shortTermAvgIndex := reShortTermAvg.SubexpIndex("shortTermAvg")
+	longTermAvgIndex := reLongTermAvg.SubexpIndex("longTermAvg")
+
+	for _, match := range reStatisticsBlock.FindAllStringSubmatch(output, -1) {
+		sws := stopwatchStatistics{}
+
+		if totalSamplesMatch := reTotalSamples.FindStringSubmatch(match[0]); totalSamplesMatch != nil {
+			sws.totalSamples = totalSamplesMatch[totalSamplesIndex]
+		}
+		if minMatch := reMinumum.FindStringSubmatch(match[0]); minMatch != nil {
+			sws.min = minMatch[minIndex]
+		}
+		if maxMatch := reMaximum.FindStringSubmatch(match[0]); maxMatch != nil {
+			sws.max = maxMatch[maxIndex]
+		}
+		if percentile95thMatch := rePercentile95th.FindStringSubmatch(match[0]); percentile95thMatch != nil {
+			sws.percentile95th = percentile95thMatch[percentile95thIndex]
+		}
+		if shortTermAvgMatch := reShortTermAvg.FindStringSubmatch(match[0]); shortTermAvgMatch != nil {
+			sws.shortTermAvg = shortTermAvgMatch[shortTermAvgIndex]
+		}
+		if longTermAvgMatch := reLongTermAvg.FindStringSubmatch(match[0]); longTermAvgMatch != nil {
+			sws.longTermAvg = longTermAvgMatch[longTermAvgIndex]
+		}
+
+		metricName := match[nameIndex]
+		result[metricName] = sws
+	}
+
+	return result
+}
+
+// stopwatchShowMetricsUpdater updates the metric by obtaining the stopwatch/show
+// metrics for the specified component.
+func stopwatchShowMetricsUpdater(component string) {
+	for {
+		time.Sleep(30 * time.Second)
+
+		stopwatchShowOutputMap, err := getStopwatchShowOutputMap(component)
+		if err != nil {
+			klog.Error(err)
+			continue
+		}
+
+		if len(stopwatchShowOutputMap) == 0 {
```

Comment:
I am not sure what the right "Prometheus-y" thing is in this case: should we set the metric to value 0, or instead skip that point in time like you are doing here? I _feel_ like we should set it to 0, but I am not sure.

Reply:
> I am not sure if I got you correctly, but this is to "inform" if there are no metrics predefined. 

By predefined here you mean: that the `ovn-appctl stopwatch/show` doesn't contain the string `Statistics for` in its output, right? Because, as I read it, that's the only reason `stopwatchShowOutputMap` can be of length 0. Is it even possible that  `ovn-appctl stopwatch/show` returns an empty output?

Reply:
Sorry, I was wrong with meaning "predefined metrics". `stopwatchShowOutputMap` contains the output from `ovn-appctl stopwatch/show`. Had a similar discussion here https://github.com/ovn-org/ovn-kubernetes/pull/2443#discussion_r706347937. The reason why I check for this and `continue` is to not set the metrics values to `0`, if there were no metrics provided.